File: packages/host4_flutter_aivoice/ios/AiVoice_副本/Http/AgentRequestManager.m
//
//  AgentRequestManager.m
//  GameMacro
//
//  Created by 敬攀 on 2025/12/11.
//

#import "AgentRequestManager.h"
#import "AFNetworking.h"
#import "NSDictionary+NullHandle.h"
#import "WYProgressHUD.h"
#import "WYImageResetSize.h"
#import "DYRequestModel.h"
#import "UIImage+DYCategory.h"
#import "MyAFHTTPSessionManager.h"

@implementation AgentRequestManager

+(void)agentJoinRoomRequest:(NSString *)boostingTableID roomID:(NSString *)roomID taskID:(NSString *)taskID userID:(NSString *)userID  botname:(NSString *)botname completedBlock:(void (^)(void))successblock failuerBlock:(void (^)(void))failblock{
    
    if (isStringEmpty(boostingTableID) || isStringEmpty(roomID) || isStringEmpty(taskID) || isStringEmpty(userID)){
        if (failblock){
            failblock();
        }
        return;
    }
    
    NSMutableDictionary *param = [[NSMutableDictionary alloc] initWithDictionary:@{@"BoostingTableID":boostingTableID,@"RoomID":roomID,@"TaskID":taskID,@"TargetUserID":userID,@"SystemLanguage":[[HXLanguageManager shareInstance] getAILanguage]}];
    
    if (!isStringEmpty(botname)){
        [param setValue:botname forKey:@"Botname"];
    }
    
    [self PostRequestWithInfo:startVoiceChat andParameters:param completedBlock:^(id data) {
        
        if (successblock){
            successblock();
        }

    } failuerBlock:^(id ErrorModal) {
        if (failblock){
            failblock();
        }
    }];
}

+(void)agentLeaveRoomRequest:(NSString *)roomID taskID:(NSString *)taskID  completedBlock:(void (^)(void))successblock failuerBlock:(void (^)(void))failblock{
    
    if (isStringEmpty(roomID) || isStringEmpty(taskID)){
        if (failblock){
            failblock();
        }
        return;
    }
    
    NSMutableDictionary *param = [[NSMutableDictionary alloc] initWithDictionary:@{@"RoomID":roomID,@"TaskID":taskID}];
    
    [self PostRequestWithInfo:stopVoiceChat andParameters:param completedBlock:^(id data) {
        
        if (successblock){
            successblock();
        }
        
    } failuerBlock:^(id ErrorModal) {
        if (failblock){
            failblock();
        }
    }];
}


+(void)agentUpdateRoomRequest:(NSString *)roomID taskID:(NSString *)taskID command:(NSString *)command  completedBlock:(void (^)(void))successblock failuerBlock:(void (^)(void))failblock{
    
    if (isStringEmpty(roomID) || isStringEmpty(taskID) || isStringEmpty(command)){
        if (failblock){
            failblock();
        }
        return;
    }
    
    NSMutableDictionary *param = [[NSMutableDictionary alloc] initWithDictionary:@{@"RoomID":roomID,@"TaskID":taskID,@"Command":command}];
    
    [self PostRequestWithInfo:updateVoiceChat andParameters:param completedBlock:^(id data) {
        
        if (successblock){
            successblock();
        }
        
    } failuerBlock:^(id ErrorModal) {
        if (failblock){
            failblock();
        }
    }];
}

+(void)PostRequestWithInfo:(NSString *)str andParameters:(NSDictionary *)parameters completedBlock:(void (^)(id))successblock failuerBlock:(void (^)(id))failblock{
    
    if (![UIActivityIndicatorView appearanceWhenContainedInInstancesOfClasses:@[[MBProgressHUD class]]]){
        [MBProgressHUD showHintTips:@"" detailStr:kLocalizedString(@"Loading", @"") toView:KEY_WINDOW];
    }
    
    AFHTTPSessionManager *manager = [self managerWithTimeOut:20];

    NSMutableDictionary *mutableDic = [[NSMutableDictionary alloc] initWithDictionary:parameters];
    
    [manager.requestSerializer setValue:@"application/json" forHTTPHeaderField:BaseContentType];
    
    NSString *urlStr = [NSString stringWithFormat:@"%@/%@",AgentChatBaseURL,str];
    
    [manager POST:urlStr parameters:mutableDic headers:nil progress:^(NSProgress * _Nonnull uploadProgress) {
        
    } success:^(NSURLSessionDataTask * _Nonnull task, id  _Nullable responseObject) {
        
        if (successblock){
            successblock(responseObject);
        }
    } failure:^(NSURLSessionDataTask * _Nullable task, NSError * _Nonnull error) {
        
        if (failblock) {
            failblock(error);
        }
    }];
}


+(AFHTTPSessionManager *)managerWithTimeOut:(float)timeOut{
    AFHTTPSessionManager * manager =  [MyAFHTTPSessionManager sharedHTTPSession];
    manager.requestSerializer.timeoutInterval = timeOut;
    return manager;
}

#pragma mark - 会员使用
+(void)requestVipUserUse:(NSInteger)type completedBlock:(void (^)(id))successblock failuerBlock:(void (^)(id))failblock{
    
    NSDictionary *dic = @{@"type":@(type)};
    
    [HttpRequrst GetRequestWithInfo:aiExpiredUserUse andParameters:dic completedBlock:^(id data) {
        
        UserVipModel *vipM = [UserVipModel mj_objectWithKeyValues:data];
        
        if (successblock){
            successblock(vipM);
        }
        
    } failuerBlock:^(id ErrorModal) {
        
        if (failblock) {
            failblock(ErrorModal);
        }
        
    }];
}


#pragma mark - 下单支付
+(void)requestCreatePayOrder:(int)setMealId completedBlock:(void (^)(id _Nonnull))successblock failuerBlock:(void (^)(id _Nonnull))failblock{
    
    NSDictionary *dic = @{@"setMealId":@(setMealId),@"payWayType":@"appleStore"};
    
    [HttpRequrst GetRequestWithInfo:aiOrderPay andParameters:dic completedBlock:^(id data) {
        
        if (successblock){
            successblock(data);
        }
        
    } failuerBlock:^(id ErrorModal) {
        
        if (failblock) {
            failblock(ErrorModal);
        }
        
    }];
}

#pragma mark - 苹果支付回调
+(void)requestAppStoreNotify:(NSString *)signedPayload completedBlock:(void (^)(id _Nonnull))successblock failuerBlock:(void (^)(id _Nonnull))failblock{
    
    NSDictionary *dic = @{@"signedPayload":signedPayload};
    
    [HttpRequrst PostRequestWithInfo:aiAppleStorePayNotify andParameters:dic completedBlock:^(id data) {
        
        BOOL isSuccess = [data boolValue];
        
        if (successblock){
            successblock(@(isSuccess));
        }
        
    } failuerBlock:^(id ErrorModal) {
        
        if (failblock) {
            failblock(ErrorModal);
        }
        
    }];
    
}


#pragma mark - 用户游戏时长记录
+(void)requestUserAddGameTime:(NSInteger)type tag:(NSString *)tag completedBlock:(void (^)(id))successblock failuerBlock:(void (^)(id))failblock{
    
    if (![DYUserManager shareManager].model){
        return;
    }
    
    NSDictionary *dic = @{@"type":@(type),@"tag":tag};
    
    [HttpRequrst PostRequestWithInfo:aiExpiredAddGameTime andParameters:dic completedBlock:^(id data) {
        
        BOOL isSuccess = [data boolValue];
        
        if (successblock){
            successblock(@(isSuccess));
        }
        
    } failuerBlock:^(id ErrorModal) {
        
        if (failblock) {
            failblock(ErrorModal);
        }
        
    }];
    
}

@end

File: packages/host4_flutter_aivoice/ios/Classes/AgentRequestManager.swift
import Foundation

/// AI 智能体 HTTP 请求管理
class AgentRequestManager {
  /// AI 服务基础 URL
  static let baseURL = "https://gamemacro.cn/ai"

  /// 智能体加入房间
  /// - Parameters:
  ///   - boostingTableID: 智能体配置 ID
  ///   - roomID: RTC 房间 ID
  ///   - taskID: 任务 ID
  ///   - userID: 用户 ID
  ///   - botname: 智能体名称（可选）
  ///   - language: 语言代码，如 "zh"
  ///   - completion: 回调（成功/失败）
  static func agentJoinRoom(
    boostingTableID: String,
    roomID: String,
    taskID: String,
    userID: String,
    botname: String?,
    language: String,
    completion: @escaping (Bool) -> Void
  ) {
    guard !boostingTableID.isEmpty, !roomID.isEmpty, !taskID.isEmpty, !userID.isEmpty else {
      print("[AgentRequest] ❌ 参数不完整")
      completion(false)
      return
    }

    var params: [String: Any] = [
      "BoostingTableID": boostingTableID,
      "RoomID": roomID,
      "TaskID": taskID,
      "TargetUserID": userID,
      "SystemLanguage": language,
    ]
    if let botname = botname, !botname.isEmpty {
      params["Botname"] = botname
    }

    post(path: "start_voice_chat", parameters: params) { success, _ in
      if success {
        print("[AgentRequest] ✅ 智能体加入房间成功")
      } else {
        print("[AgentRequest] ❌ 智能体加入房间失败")
      }
      completion(success)
    }
  }

  /// 智能体退出房间
  static func agentLeaveRoom(
    roomID: String,
    taskID: String,
    completion: @escaping (Bool) -> Void
  ) {
    guard !roomID.isEmpty, !taskID.isEmpty else {
      completion(false)
      return
    }
    let params: [String: Any] = ["RoomID": roomID, "TaskID": taskID]
    post(path: "stop_voice_chat", parameters: params) { success, _ in
      completion(success)
    }
  }

  /// 智能体更新
  static func agentUpdateRoom(
    roomID: String,
    taskID: String,
    command: String,
    completion: @escaping (Bool) -> Void
  ) {
    guard !roomID.isEmpty, !taskID.isEmpty, !command.isEmpty else {
      completion(false)
      return
    }
    let params: [String: Any] = ["RoomID": roomID, "TaskID": taskID, "Command": command]
    post(path: "update_voice_chat", parameters: params) { success, _ in
      completion(success)
    }
  }

  // MARK: - Private

  private static func post(
    path: String,
    parameters: [String: Any],
    completion: @escaping (Bool, Any?) -> Void
  ) {
    guard let url = URL(string: "\(baseURL)/\(path)") else {
      completion(false, nil)
      return
    }

    var request = URLRequest(url: url)
    request.httpMethod = "POST"
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    request.timeoutInterval = 20

    do {
      request.httpBody = try JSONSerialization.data(withJSONObject: parameters)
    } catch {
      completion(false, nil)
      return
    }

    URLSession.shared.dataTask(with: request) { data, response, error in
      if let error = error {
        print("[AgentRequest] ❌ \(path) 失败: \(error.localizedDescription)")
        DispatchQueue.main.async { completion(false, error) }
        return
      }
      if let httpResponse = response as? HTTPURLResponse {
        print("[AgentRequest] \(path) status: \(httpResponse.statusCode)")
      }
      DispatchQueue.main.async { completion(true, data) }
    }.resume()
  }
}
